### PR TITLE
Fix overflow error in shuffle colormap for signed integer labels

### LINKED ISSUE #7277: OverflowError when shuffling colors of labeled image when having two Labels Layers
### 🐛 Bug Report

I generated two 3D boolean masks. If I move in Z and then shuffle the colors of one Labels Layer, I get an OverflowError.

### 💡 Steps to Reproduce

Open Napari

Run the following on the terminal

```
import napari
from skimage.morphology import ball

mask_1 = ball(6).astype(bool)
mask_2 = mask_1.copy()
mask_2[4:8, 4:8, 4:8] = 0

viewer = napari.Viewer()
viewer.add_labels(mask_1)
viewer.add_labels(mask_2)

napari.run()
```

Move to any Z-plane that is not the initial one.
Shuffle colors in one of the Labels Layer.

You get an OverflowError

### 💡 Expected Behavior

I expected the colors of the Labels to change and Napari to still work.

### 🌎 Environment

Python: 3.12.5
napari: 0.5.3
vispy: 0.14.3

### 💡 Additional Context

Pasted Error:

<details>

<summary>

```pytb
File ~\.conda\envs\ccm_pipe\Lib\site-packages\napari\utils\colormaps\_accelerated_cmap.py:77, in zero_preserving_modulo_numpy(values=array([ 0,  1, ..., -2, -1], dtype=int8), n=228, dtype=dtype('int8'), to_zero=0)
     51 def zero_preserving_modulo_numpy(
     52     values: np.ndarray, n: int, dtype: np.dtype, to_zero: int = 0
     53 ) -> np.ndarray:
     54     """``(values - 1) % n + 1``, but with one specific value mapped to 0.
     55
     56     This ensures (1) an output value in [0, n] (inclusive), and (2) that
   (...)
     75         everywhere else.
     76     """
---> 77     res = ((values - 1) % n + 1).astype(dtype)
        dtype = dtype('int8')
        values = array([ 0,  1, ..., -2, -1], dtype=int8)
        n = 228
     78     res[values == to_zero] = 0
     79     return res

OverflowError: Python integer 228 out of bounds for int8
```

</summary>

```pytb
OverflowError                             Traceback (most recent call last)
File ~\.conda\envs\ccm_pipe\Lib\site-packages\napari\_qt\layer_controls\qt_labels_controls.py:375, in QtLabelsControls.changeColor(self=<napari._qt.layer_controls.qt_labels_controls.QtLabelsControls object>)
    373 def changeColor(self):
    374     """Change colormap of the label layer."""
--> 375     self.layer.new_colormap()
        self.layer = <Labels layer 'mask_2' at 0x15b534f3950>
        self = <napari._qt.layer_controls.qt_labels_controls.QtLabelsControls object at 0x0000015B45B46F90>

File ~\.conda\envs\ccm_pipe\Lib\site-packages\napari\layers\labels\labels.py:523, in Labels.new_colormap(self=<Labels layer 'mask_2'>, seed=np.int64(372634340))
    520     seed = np.random.default_rng().integers(2**32 - 1)
    522 orig = self._original_random_colormap
--> 523 self.colormap = shuffle_and_extend_colormap(
        self = <Labels layer 'mask_2' at 0x15b534f3950>
        self._original_random_colormap = CyclicLabelColormap(colors=ColorArray([[0.61961, 0.52941, 0.43529, 1.     ],
            [0.35686, 0.83529, 0.97255, 1.     ],
            ...,
            [0.23529, 0.10588, 0.52941, 1.     ],
            [0.84314, 0.58431, 0.33333, 1.     ]], dtype=float32), name='label_colormap', interpolation=<ColormapInterpolationMode.ZERO: 'zero'>, controls=array([0.     , 0.00437, ..., 0.99563, 1.     ]), use_selection=False, selection=0, background_value=0, seed=0.5)
        seed = np.int64(372634340)
    524     self._original_random_colormap, seed
    525 )
    526 self._original_random_colormap = orig

File ~\.conda\envs\ccm_pipe\Lib\site-packages\napari\layers\labels\labels.py:534, in Labels.colormap(self=<Labels layer 'mask_2'>, colormap=CyclicLabelColormap(colors=ColorArray([[0.61961,...False, selection=0, background_value=0, seed=0.5))
    532 @colormap.setter
    533 def colormap(self, colormap: LabelColormapBase):
--> 534     self._set_colormap(colormap)
        colormap = CyclicLabelColormap(colors=ColorArray([[0.61961, 0.52941, 0.43529, 1.     ],
            [0.35686, 0.83529, 0.97255, 1.     ],
            ...,
            [0.23529, 0.10588, 0.52941, 1.     ],
            [0.84314, 0.58431, 0.33333, 1.     ]], dtype=float32), name='label_colormap', interpolation=<ColormapInterpolationMode.ZERO: 'zero'>, controls=array([0.     , 0.00437, ..., 0.99563, 1.     ]), use_selection=False, selection=0, background_value=0, seed=0.5)
        self = <Labels layer 'mask_2' at 0x15b534f3950>

File ~\.conda\envs\ccm_pipe\Lib\site-packages\napari\layers\labels\labels.py:562, in Labels._set_colormap(self=<Labels layer 'mask_2'>, colormap=CyclicLabelColormap(colors=ColorArray([[0.61961,...False, selection=0, background_value=0, seed=0.5))
    560 self._selected_color = self.get_color(self.selected_label)
    561 self._color_mode = color_mode
--> 562 self.events.colormap()  # Will update the LabelVispyColormap shader
        self = <Labels layer 'mask_2' at 0x15b534f3950>
        self.events.colormap = <napari.utils.events.event.EventEmitter object at 0x0000015B55CD4470>
        self.events = <napari.utils.events.event.EmitterGroup object at 0x0000015B55F1C2C0>
    563 self.events.selected_label()
    564 self.refresh(extent=False)

File ~\.conda\envs\ccm_pipe\Lib\site-packages\napari\utils\events\event.py:764, in EventEmitter.__call__(self=<napari.utils.events.event.EventEmitter object>, *args=(), **kwargs={})
    761     self._block_counter.update([cb])
    762     continue
--> 764 self._invoke_callback(cb, event if pass_event else None)
        event = <Event blocked=False handled=False native=None source=None sources=[] type='colormap'>
        self = <napari.utils.events.event.EventEmitter object at 0x0000015B55CD4470>
        cb = <bound method VispyLabelsLayer._on_colormap_change of <napari._vispy.layers.labels.VispyLabelsLayer object at 0x0000015B55ED72F0>>
        pass_event = True
    765 if event.blocked:
    766     break

File ~\.conda\envs\ccm_pipe\Lib\site-packages\napari\utils\events\event.py:802, in EventEmitter._invoke_callback(self=<napari.utils.events.event.EventEmitter object>, cb=<bound method VispyLabelsLayer._on_colormap_chan...ri._vispy.layers.labels.VispyLabelsLayer object>>, event=<Event blocked=False handled=False native=None source=None sources=[] type='colormap'>)
    800     self.disconnect(cb)
    801     return
--> 802 _handle_exception(
        self = <napari.utils.events.event.EventEmitter object at 0x0000015B55CD4470>
        event = <Event blocked=False handled=False native=None source=None sources=[] type='colormap'>
        cb = <bound method VispyLabelsLayer._on_colormap_change of <napari._vispy.layers.labels.VispyLabelsLayer object at 0x0000015B55ED72F0>>
        (cb, event) = (<bound method VispyLabelsLayer._on_colormap_change of <napari._vispy.layers.labels.VispyLabelsLayer object at 0x0000015B55ED72F0>>, <Event blocked=False handled=False native=None source=None sources=[] type='colormap'>)
    803     self.ignore_callback_errors,
    804     self.print_callback_errors,
    805     self,
    806     cb_event=(cb, event),
    807 )

File ~\.conda\envs\ccm_pipe\Lib\site-packages\napari\utils\events\event.py:789, in EventEmitter._invoke_callback(self=<napari.utils.events.event.EventEmitter object>, cb=<bound method VispyLabelsLayer._on_colormap_chan...ri._vispy.layers.labels.VispyLabelsLayer object>>, event=<Event blocked=False handled=False native=None source=None sources=[] type='colormap'>)
    787 try:
    788     if event is not None:
--> 789         cb(event)
        event = <Event blocked=False handled=False native=None source=None sources=[] type='colormap'>
        cb = <bound method VispyLabelsLayer._on_colormap_change of <napari._vispy.layers.labels.VispyLabelsLayer object at 0x0000015B55ED72F0>>
    790     else:
    791         cb()

File ~\.conda\envs\ccm_pipe\Lib\site-packages\napari\_vispy\layers\labels.py:254, in VispyLabelsLayer._on_colormap_change(self=<napari._vispy.layers.labels.VispyLabelsLayer object>, event=<Event blocked=False handled=False native=None source=None sources=[] type='colormap'>)
    250     colormap = CyclicLabelColormap(**colormap.dict())
    251     colormap.background_value = (
    252         colormap._background_as_minimum_dtype(raw_dtype)
    253     )
--> 254 color_texture = _select_colormap_texture(
        colormap = CyclicLabelColormap(colors=ColorArray([[0.61961, 0.52941, 0.43529, 1.     ],
            [0.35686, 0.83529, 0.97255, 1.     ],
            ...,
            [0.23529, 0.10588, 0.52941, 1.     ],
            [0.84314, 0.58431, 0.33333, 1.     ]], dtype=float32), name='label_colormap', interpolation=<ColormapInterpolationMode.ZERO: 'zero'>, controls=array([0.     , 0.00437, ..., 0.99563, 1.     ]), use_selection=False, selection=0, background_value=0, seed=0.5)
        view_dtype = dtype('uint8')
        raw_dtype = dtype('int8')
    255     colormap, view_dtype, raw_dtype
    256 )
    257 self.node.cmap = LabelVispyColormap(
    258     colormap, view_dtype=view_dtype, raw_dtype=raw_dtype
    259 )
    260 self.node.shared_program['texture2D_values'] = Texture2D(
    261     color_texture,
    262     internalformat='rgba32f',
    263     interpolation='nearest',
    264 )

File ~\.conda\envs\ccm_pipe\Lib\site-packages\napari\_vispy\layers\labels.py:186, in _select_colormap_texture(colormap=CyclicLabelColormap(colors=ColorArray([[0.61961,...False, selection=0, background_value=0, seed=0.5), view_dtype=dtype('uint8'), raw_dtype=dtype('int8'))
    184     color_texture = colormap._get_mapping_from_cache(view_dtype)
    185 else:
--> 186     color_texture = colormap._get_mapping_from_cache(raw_dtype)
        raw_dtype = dtype('int8')
        colormap = CyclicLabelColormap(colors=ColorArray([[0.61961, 0.52941, 0.43529, 1.     ],
            [0.35686, 0.83529, 0.97255, 1.     ],
            ...,
            [0.23529, 0.10588, 0.52941, 1.     ],
            [0.84314, 0.58431, 0.33333, 1.     ]], dtype=float32), name='label_colormap', interpolation=<ColormapInterpolationMode.ZERO: 'zero'>, controls=array([0.     , 0.00437, ..., 0.99563, 1.     ]), use_selection=False, selection=0, background_value=0, seed=0.5)
    188 if color_texture is None:
    189     raise ValueError(  # pragma: no cover
    190         f'Cannot build a texture for dtype {raw_dtype=} and {view_dtype=}'
    191     )

File ~\.conda\envs\ccm_pipe\Lib\site-packages\napari\utils\colormaps\colormap.py:221, in LabelColormapBase._get_mapping_from_cache(self=CyclicLabelColormap(colors=ColorArray([[0.61961,...False, selection=0, background_value=0, seed=0.5), data_dtype=dtype('int8'))
    217 if key not in self._cache_mapping and data_dtype.itemsize <= 2:
    218     data = np.arange(
    219         np.iinfo(target_dtype).max + 1, dtype=target_dtype
    220     ).astype(data_dtype)
--> 221     self._cache_mapping[key] = self._map_without_cache(data)
        key = (dtype('int8'), dtype('uint8'))
        data = array([ 0,  1, ..., -2, -1], dtype=int8)
        self = CyclicLabelColormap(colors=ColorArray([[0.61961, 0.52941, 0.43529, 1.     ],
            [0.35686, 0.83529, 0.97255, 1.     ],
            ...,
            [0.23529, 0.10588, 0.52941, 1.     ],
            [0.84314, 0.58431, 0.33333, 1.     ]], dtype=float32), name='label_colormap', interpolation=<ColormapInterpolationMode.ZERO: 'zero'>, controls=array([0.     , 0.00437, ..., 0.99563, 1.     ]), use_selection=False, selection=0, background_value=0, seed=0.5)
        self._cache_mapping = {}
    222 return self._cache_mapping.get(key)

File ~\.conda\envs\ccm_pipe\Lib\site-packages\napari\utils\colormaps\colormap.py:316, in CyclicLabelColormap._map_without_cache(self=CyclicLabelColormap(colors=ColorArray([[0.61961,...False, selection=0, background_value=0, seed=0.5), values=array([ 0,  1, ..., -2, -1], dtype=int8))
    315 def _map_without_cache(self, values) -> np.ndarray:
--> 316     texture_dtype_values = _accel_cmap.zero_preserving_modulo_numpy(
        _accel_cmap = <module 'napari.utils.colormaps._accelerated_cmap' from 'C:\\Users\\aguco599\\.conda\\envs\\ccm_pipe\\Lib\\site-packages\\napari\\utils\\colormaps\\_accelerated_cmap.py'>
        values = array([ 0,  1, ..., -2, -1], dtype=int8)
        self.colors = ColorArray([[0.61961, 0.52941, 0.43529, 1.     ],
            [0.35686, 0.83529, 0.97255, 1.     ],
            ...,
            [0.23529, 0.10588, 0.52941, 1.     ],
            [0.84314, 0.58431, 0.33333, 1.     ]], dtype=float32)
        self = CyclicLabelColormap(colors=ColorArray([[0.61961, 0.52941, 0.43529, 1.     ],
            [0.35686, 0.83529, 0.97255, 1.     ],
            ...,
            [0.23529, 0.10588, 0.52941, 1.     ],
            [0.84314, 0.58431, 0.33333, 1.     ]], dtype=float32), name='label_colormap', interpolation=<ColormapInterpolationMode.ZERO: 'zero'>, controls=array([0.     , 0.00437, ..., 0.99563, 1.     ]), use_selection=False, selection=0, background_value=0, seed=0.5)
        self.background_value = 0
    317         values,
    318         len(self.colors) - 1,
    319         values.dtype,
    320         self.background_value,
    321     )
    322     mapped = self.colors[texture_dtype_values]
    323     mapped[texture_dtype_values == 0] = 0

File ~\.conda\envs\ccm_pipe\Lib\site-packages\napari\utils\colormaps\_accelerated_cmap.py:77, in zero_preserving_modulo_numpy(values=array([ 0,  1, ..., -2, -1], dtype=int8), n=228, dtype=dtype('int8'), to_zero=0)
     51 def zero_preserving_modulo_numpy(
     52     values: np.ndarray, n: int, dtype: np.dtype, to_zero: int = 0
     53 ) -> np.ndarray:
     54     """``(values - 1) % n + 1``, but with one specific value mapped to 0.
     55
     56     This ensures (1) an output value in [0, n] (inclusive), and (2) that
   (...)
     75         everywhere else.
     76     """
---> 77     res = ((values - 1) % n + 1).astype(dtype)
        dtype = dtype('int8')
        values = array([ 0,  1, ..., -2, -1], dtype=int8)
        n = 228
     78     res[values == to_zero] = 0
     79     return res

OverflowError: Python integer 228 out of bounds for int8
```

</details>

### PR DESCRIPTION
**Pull request by Czaki**
# References and relevant issues

closes #7277

# Description

Fix `zero_preserving_modulo_numpy` by not performing modulo operation when `n` is bigger than dtype of `values`

### CODE REVIEW

**Comment by Czaki** [comment]
Another option could be to divide n/2 if data are signed integers.

**Comment by jni** [comment]
@Czaki this doesn't close #7294 (I've removed that from the description). That PR is still worth merging because it fixes the issue of painting into the bool volume. I also think np.uint8 is preferable anyway, just because it's more commonly used.

And something feels a bit weird about this solution (though I'm still happy to merge it because it fixes the crash): we are relying on negative indexing into the n colours. I had to think about it, and indeed it checks out that min of the dtype, usually np.iinfo(dtype).max - 1, is guaranteed to fit in n, given the checks. Another non-obvious thing: indexing (-k) into an array of size n is the same as indexing with (-k) % n, if precision is not a concern. That's nice because we want the mapping to a color not to change between data types.

So I like this solution :+1: , but it does lead to some mental gymnastics, and I think it's better to avoid them in the simple case of bools. So all that to say I think we should merge this *and* #7294.

**Comment by jni** [comment]
I'll try to write a test and push

**Review by jni — APPROVED** [review]

**Comment by jni** [comment]
Boom. This is ready. I'll merge shortly and include in the release! 🚀

File: napari/_vispy/_tests/test_vispy_labels_layer.py [hunk 1]
```diff
@@ -126,3 +126,13 @@ def test_labels_iso_gradient_modes(qtbot, qt_viewer):
...
     assert visual.node.iso_gradient_mode == 'fast'
+
+
+def test_int8_labels_color(qtbot, qt_viewer):
+    """Check that int8 labels colors can be shuffled without overflow.
+
+    See https://github.com/napari/napari/issues/7277.
+    """
+    data = np.arange(-128, 128, dtype=np.int8).reshape((16, 16))
+    layer = qt_viewer.viewer.add_labels(data)
```
Comment (Czaki) [inline napari/_vispy/_tests/test_vispy_labels_layer.py]: Why qt_viewer, why not just create Labels layer?

Comment (jni) [reply napari/_vispy/_tests/test_vispy_labels_layer.py]: good point! I wasn't 100% sure whether I needed VispyLabels to get instantiated to see the effect, but I just checked and I can indeed cause the failure with just labels. Will push a smaller test in test_labels.py

Comment (jni) [reply napari/_vispy/_tests/test_vispy_labels_layer.py]: Fixed now. Will merge if CI passes. (new test passed locally, and fails in main)

**Comment by jni** [comment]
Oh whew, the test failure is just a failure with pinging GH urls 😅 

https://github.com/napari/napari/actions/runs/11048379513/job/30692381375?pr=7296#step:13:855

```
  =================================== FAILURES ===================================
  _________________________ test_help_urls[github_issue] _________________________
```

I'm going to ignore it. Merging!

**Pull request merged**